clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -200,6 +200,102 @@ bool Sema::isKnownGoodSYCLDecl(const Decl *D) {
   return false;
 }
 
+static bool isZeroSizedArray(QualType Ty) {
+  if (const auto *CATy = dyn_cast<ConstantArrayType>(Ty))
+    return CATy->getSize() == 0;
+  return false;
+}
+
+static Sema::DeviceDiagBuilder
+emitDeferredDiagnosticAndNote(Sema &S, SourceRange Loc, unsigned DiagID,
+                              SourceRange UsedAtLoc) {
+  Sema::DeviceDiagBuilder builder =
+      S.SYCLDiagIfDeviceCode(Loc.getBegin(), DiagID);
+  if (UsedAtLoc.isValid())
+    S.SYCLDiagIfDeviceCode(UsedAtLoc.getBegin(), diag::note_sycl_used_here);
+  return builder;
+}
+
+static void checkSYCLVarType(Sema &S, QualType Ty, SourceRange Loc,
+                             llvm::DenseSet<QualType> Visited,
+                             SourceRange UsedAtLoc = SourceRange()) {
+  // Not all variable types are supported inside SYCL kernels,
+  // for example the quad type __float128 will cause the resulting
+  // SPIR-V to not link.
```

Comment:
> __float128 will cause the resulting SPIR-V to not link.

What exactly do you mean here? AFAIK we don't link SPIR-V files.

Reply:
I am not an expert on the pipeline, especially in regards to how SPIR-V is handled. In the simplest case, without these type checks, if you have kernel code that contains an unsupported type (such as __int128) , you can pass the -c flag to the compiler to simply output the object files, and everything is fine. But when you next go to link those files into your executable, you'll get an error ( `InvalidBitWidth: Invalid bit width in input: 128`  etc) .  IIRC, those are emitted by SpirVWriter.cpp 
So it may not be linking, per se. But it occurs in what we normally think of as the link phase.

Reply:
Work with @Fznamznon  to make sure that this comment means what you want it to.

Reply:
Hmm okay, such errors actually happen during LLVM IR -> SPIR-V translation and it really happen somewhere near the link phase, but technically it is not linking.
Let's do not confuse anyone and mention something like that: "unsupported types in device code cause errors on SPIR-V translation stage." You can rewrite it in a better way, but keep it true.

Reply:
updated it